csrc/scheduler/normalization.cpp

```diff
@@ -124,20 +124,30 @@ std::shared_ptr<ReductionParams> innerOuterPersistentHeuristic(
   // inner_dim_numel is dividable by the multiplication of a quarter warp and
   // vectorize_factor.
   iop.inner_vect = (int64_t)vectorize_factor;
-  auto opt_inner_batch = normalization_scheduler_utils::
+
+  // Enforce valid return is needed because we enforced projection for fp32 if
```

Comment:
Could you elaborate more why the register constraint needs to be ignored? `getOptionalInnerOuterPersistentBufferBatches` is also used by `canScheduleRuntime`, so doesn't that mean the fusion is guaranteed to pass the register threshold?

Reply:
The register usage in `canScheduleRuntime` is based on std::min(project_buffer, not_project_buffer). However, in scheduler we enforce project_buffer to input if dtype=float and feature size <=14K. It leads to register spills but still faster than unprojected version due to the reuse of a input para in this grid persistent kernel.
This is a tmp solution before we have a new persistent heuristics, where the projection is not soley based on size of buffers.

Reply:
updated comments.

Reply:
So, the value of `max_persistent_buffer_size` here is different from the one used in `canScheduleRuntime`?

Reply:
Yes, they are different. if dtype is float and feature size is less or equal than 14K.

Reply:
extended comments:
```
  // ignore_register_size_limit will return a valid batch size.
  // This is needed because we enforced projection for fp32 if the feature size
  // is less or equal 14K. It leads to register spills but still faster than the
  // unprojected version due to the reuse of a input para in this grid
  // persistent kernel. This is a tmp solution before we have a new persistent
  // heuristics, where the projection is not solely based on size of buffers.
```

Reply:
Thanks. One of the reasons this part looks confusing to me is that there's the same check in `canScheduleRuntime`. The comment seems to be missing some important facts like:

- The enforced buffer projection is not considered in `canScheduleRuntime`
- Thus, `max_persistent_buffer_size` is larger than the value used in `canScheduleRuntime`.

This may sound obvious to you, but I'm pretty sure it isn't the case unless you have a complete understanding of the `canSchedule` checks as well as the actual scheduling logic.

Reply:
Thanks for the suggestions. The comment is further revised.
```
  // ignore_register_size_limit will return a valid batch size.
  // This is needed because we enforced projection for fp32 if the feature size
  // is less or equal 14K. It leads to register spills but still faster than the
  // unprojected version due to the reuse of a input para in this grid
  // persistent kernel. However, when we do register usage check in
  // canScheduleRuntime, the enforced projection is not considered. Thus,
  // max_persistent_buffer_size used here is larger than the value used in
  // canScheduleRuntime.
  // This is a tmp solution before we have a new persistent heuristics, where
  // the projection is not solely based on size of buffers. The enforced buffer
  // projection is not considered in canScheduleRuntime Thus,
```